units/migrate_volume_job.go

```diff
@@ -0,0 +1,298 @@
+}
+
+type volumeMigrationJob struct {
+	job.Base      `bson:"job_base" json:"job_base" yaml:"job_base"`
+	VolumeID      string             `bson:"volume_id" yaml:"volume_id"`
+	ModifyOptions cloud.SpawnOptions `bson:"modify_options" json:"modify_options" yaml:"modify_options"`
+
+	VolumeDetached        bool   `bson:"volume_detached"`
+	InitialHostID         string `bson:"initial_host_id"`
+	SpawnhostStopJobID    string `bson:"spawnhost_stop_job_id"`
+	InitialHostStopped    bool   `bson:"initial_host_stopped"`
+	InitialHostTerminated bool   `bson:"initial_host_terminated"`
+	NewHostID             string `bson:"new_host_id"`
+
+	env         evergreen.Environment
+	volume      *host.Volume
+	initialHost *host.Host
+}
+
+func makeVolumeMigrationJob() *volumeMigrationJob {
+	j := &volumeMigrationJob{
+		Base: job.Base{
+			JobType: amboy.JobType{
+				Name:    volumeMigrationJobName,
+				Version: 0,
+			},
+		},
+	}
+	return j
+}
+
+func NewVolumeMigrationJob(env evergreen.Environment, volumeID string, modifyOptions cloud.SpawnOptions, ts string) amboy.Job {
+	j := makeVolumeMigrationJob()
+	j.env = env
+	j.SetID(fmt.Sprintf("%s.%s.%s", volumeMigrationJobName, volumeID, ts))
+	j.SetScopes([]string{fmt.Sprintf("%s.%s", volumeMigrationJobName, volumeID)})
+	j.SetEnqueueAllScopes(true)
+	j.VolumeID = volumeID
+	j.ModifyOptions = modifyOptions
+	j.UpdateRetryInfo(amboy.JobRetryOptions{
+		Retryable:   utility.TruePtr(),
+		MaxAttempts: utility.ToIntPtr(30),
+		WaitUntil:   utility.ToTimeDurationPtr(20 * time.Second),
+	})
+
+	return j
+}
+
+func (j *volumeMigrationJob) Run(ctx context.Context) {
+	defer j.finishJob(ctx)
+
+	if err := j.populateIfUnset(); err != nil {
+		j.AddRetryableError(err)
+		return
+	}
+
+	if err := j.volume.SetMigrating(true); err != nil {
+		j.AddRetryableError(err)
+		return
+	}
+
+	if !j.InitialHostStopped {
+		initialHostStopped, hasErr := j.stopInitialHost(ctx)
+		if hasErr {
+			return
+		}
+		if !initialHostStopped {
+			j.AddRetryableError(errors.Errorf("initial host '%s' not yet stopped", j.InitialHostID))
+			return
+		}
+	}
+
+	if !j.VolumeDetached {
+		// Unmount volume from initial host
+		if err := j.initialHost.UnsetHomeVolume(); err != nil {
+			j.AddError(errors.Wrapf(err, "unsetting home volume '%s' from host '%s'", j.VolumeID, j.InitialHostID))
+			return
+		}
+
+		if _, err := cloud.DetachVolume(ctx, j.VolumeID); err != nil {
+			j.AddError(errors.Wrapf(err, "detaching volume '%s'", j.VolumeID))
+			return
+		}
+		j.VolumeDetached = true
+
+		// Update in-memory volume
+		volume, err := j.getVolume()
+		if err != nil {
+			j.AddError(err)
+			return
+		}
+		j.volume = volume
+	}
+
+	// Avoid recreating new host on retry
+	if j.NewHostID == "" {
+		if hasErr := j.startNewHost(ctx); hasErr {
+			return
+		}
+	}
+
+	// Terminate initial host
+	if !j.InitialHostTerminated {
+		ts := utility.RoundPartOfMinute(1).Format(TSFormat)
+		terminateJob := NewSpawnHostTerminationJob(j.initialHost, evergreen.User, ts)
```

Comment:
I'm a little nervous about actually terminating the initial host. Is there a world in which the migration could be problematic or the image not suitable for the user or etc and they may want to go back to their old instance? (I have a fuzzy memory of discussing this on one of the docs but not sure if that was real haha). I wonder if we might instead want to just set the host to expirable in a few days or something to give them a chance to go back.

Reply:
That's a good point—I guess we have to assume anything is possible. I like the idea of setting the expiry date, or we could make it an option for users to select whether or not the host terminates?